src/cursor/cur_backup.c

```diff
@@ -833,6 +833,7 @@ __backup_start(
          * backup file list until it is complete and valid.
          */
         WT_WITH_HOTBACKUP_WRITE_LOCK(session, WT_CONN_HOTBACKUP_START(conn));
+        conn->hot_backup_timestamp = conn->txn_global.last_ckpt_timestamp;
```

Comment:
I think there a 3 threads here and I don't know the full interactions between them to define what behaviour is valid or possible. 

1. There's the checkpoint thread that can update this timestamp when a checkpoint ends
2. There's the user thread calling for a backup, which will set this field after it has taken and released the hot backup write lock, and atomically set a `hot_backup_start` variable.
3. There's another user thread querying the `hot_backup_timestamp` concurrently.

With the current implementation it is possible that `hot_backup_start` has been written to but the `hot_backup_timestamp` field is still unset or refers to the previous `hot_backup_start`. I think that would be an unsafe outcome but I can't tell for sure. Can we lock around reading and writing this field?

Also what makes the interaction between backup and checkpoint safe? They could interleave in some unhelpful way with my current understanding but I'm guessing there are other locks that prevent them from overlapping?

Reply:
The thread (2 in your list) is in `backup_start` here and holds both the CHECKPOINT and SCHEMA locks at the time it records the `last_ckpt_timestamp`. This is the same `last_ckpt_timestamp` value that the user can get querying today.

I think these statements should be swapped so that we set the timestamp before indicating that a backup is running. That way if there is a concurrent thread querying (there should not be) then it won't get an old value. The expected usage is that the same thread first opens a backup cursor and then, secondly, queries the timestamp.

Reply:
> That way if there is a concurrent thread querying (there should not be) then it won't get an old value.
I agree there should not be but I've learned that if you let higher layer in the stack do something unexpected they inevitably will. So I'm glad we can fix for that.

This PR has a different issue, where we are using locks, kind of to signal that a variable is safe to read. In this case we define an ordering between the `backup_start` variable and the `hot_backup_timestamp`. On the writer side we write the `hot_backup_timestamp` and then take a lock which gives us the ordering guarantee we need around the signal variable `backup_start`. On the read side in `txn_timestamp.c` there is the read of the signal variable but no ordering around the two reads. The atomic load may appear to be ordering but it actually isn't as it's a relaxed load.

Given we are following the P&R usage pattern I would expect to see a `WT_RELEASE_WRITE` on the writer side and a `WT_ACQUIRE_READ` on the reader side. However the intermixing of this pattern and the hot backup lock is _not_ recommended. 

For this reason I ask that we read the timestamp in the hotbackup read lock and write to it within the hot backup write lock. This will require extending the scope of the write lock beyond that of the `WT_WITH_HOTBACKUP_WRITE_LOCK` macro. This will also let us drop the atomic load on the read side. Thus write some becomes (pseudocode somewhat):
```
__wt_write_lock(hot_backup_lock);
conn->hot_backup_timestamp = conn->txn_global.last_ckpt_timestamp; // The order of these statements doesn't matter within a lock which is nice.
WT_CONN_HOTBACKUP_START(conn);
__wt_write_unlock(hot_backup_lock);
```

And the read side becomes:
```
__wt_read_lock(hot_backup_lock);
if (hot_backup_start)
    ts = conn->hot_backup_timestamp;
__wt_read_unlock(hot_backup_lock);
```

This is much easier to reason about than intermixing P&R, locks, and atomics.